compiler/sem/semstmts.nim

```diff
@@ -2944,24 +2993,151 @@ proc semConverterDef(c: PContext, n: PNode): PNode =
   addConverterDef(c, LazySym(sym: s))
 
 proc semMacroDef(c: PContext, n: PNode): PNode =
+  addInNimDebugUtils(c.config, "semMacroDef", n, result)
   checkSonsLen(n, bodyPos + 1, c.config)
-  result = semProcAux(c, n, skMacro, macroPragmas)
-  # macros can transform macros to nothing:
-  if namePos >= result.safeLen: return result
-  # bug #7093: if after a macro transformation we don't have an
-  # nkIteratorDef aynmore, return. The iterator then might have been
-  # sem'checked already. (Or not, if the macro skips it.)
-  if result.kind != nkMacroDef: return
-  var s = result[namePos].sym
-  var t = s.typ
-  var allUntyped = true
-  for i in 1..<t.n.len:
-    let param = t.n[i].sym
-    if param.typ.kind != tyUntyped: allUntyped = false
-  if allUntyped: incl(s.flags, sfAllUntyped)
-  if n[bodyPos].kind == nkEmpty:
-    localReport(c.config, n, reportSym(
-      rsemImplementationExpected, s))
+
+  result = shallowCopy(n)
```

Comment:
Because we're using `result` in a few places where `n` was previously used, I vaguely recall there were node flag(s) that didn't get copied (not in `persistentNodeFlags`) and caused some breakages.

Sorry, memory is fuzzy at the moment.

Reply:
I looked into it, and the only flags that I think might be an issues are `nfNoRewrite` and ` nfHasComment`. Given that `semTemplateDef` also doesn't copy the flags and, to my knowledge, there haven't been any issues so far, I think not copying non-persistent flags is safe here.

Nonetheless, comments are indeed no longer copied at the moment (neither for macros nor other routines), so that still needs to be fixed.

Reply:
I missed that `copyNodeImpl` copies the comment, so no additional action is required for them.